servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/BufferStrategies.java

```diff
@@ -0,0 +1,298 @@
+                        state.beforeNewAccumulatorEmitted(accum);
+                        return accum;
+                    })
+                    .repeat(__ -> true));
+                });
+            }
+
+            @Override
+            public int bufferSizeHint() {
+                return count;
+            }
+        };
+    }
+
+    private static final class State<T, B> {
+        private static final Object THRESHOLD_BREACHED_BEFORE_SUBSCRIBE = new Object();
+        @SuppressWarnings("rawtypes")
+        private static final AtomicReferenceFieldUpdater<State, Object> stateUpdater =
+                newUpdater(State.class, Object.class, "state");
+
+        /**
+         * The following values can be set to this state:
+         * <ul>
+         *     <li>{@code null} if neither an active accumulator is present nor a subscriber.</li>
+         *     <li>An {@link Accumulator} which is expected to accumulating items for the current boundary.</li>
+         *     <li>{@link AccumulatorAndSubscriber} instance when an active accumulator and {@link Subscriber} both are
+         *     active.</li>
+         *     <li>{@link Subscriber} if the size threshold {@link Completable} is subscribed and there is no
+         *     accumulator present.</li>
+         * </ul>
+         */
+        @Nullable
+        private volatile Object state;
+
+        void countThresholdBreached(final Accumulator<T, B> breachedAccumulator) {
+            for (;;) {
+                final Object cState = state;
+                if (cState == THRESHOLD_BREACHED_BEFORE_SUBSCRIBE || cState == null) {
+                    // multiple breaches
+                    return;
+                }
+                if (cState instanceof Accumulator) {
+                    if (cState != breachedAccumulator ||
+                            stateUpdater.compareAndSet(this, cState, THRESHOLD_BREACHED_BEFORE_SUBSCRIBE)) {
+                        // Either the threshold was breached for a stale accumulator or threshold was breached before
+                        // the accumulator could be set.
+                        return;
+                    }
+                } else if (cState instanceof Subscriber) {
+                    if (stateUpdater.compareAndSet(this, cState, null)) {
+                        Subscriber subscriber = (Subscriber) cState;
+                        subscriber.onComplete();
+                        return;
+                    }
+                } else if (cState instanceof AccumulatorAndSubscriber) {
+                    @SuppressWarnings("unchecked")
+                    final AccumulatorAndSubscriber<T, B> accumulatorAndSubscriber =
+                            (AccumulatorAndSubscriber<T, B>) cState;
+                    if (accumulatorAndSubscriber.accumulator != breachedAccumulator ||
+                            stateUpdater.compareAndSet(this, cState, null)) {
+                        return;
```

Comment:
if we did `stateUpdater.compareAndSet(this, cState, null)` ... shouldn't we also do `subscriber.onComplete()`?